File: algorithms/90. Subsets II/solution.cc
class Solution {
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        vector<int> path;
        vector<vector<int>> res;
        sort(nums.begin(), nums.end());
        dfs(0, nums, path, res);
        return res;
    }

    void dfs(int start, vector<int> & nums, vector<int> & path, vector<vector<int>> & res) {
        res.push_back(path);
        for ( int i = start; i < nums.size(); ++i ) {
            if ( i == start || nums[i] != nums[i-1] ) {
                path.push_back(nums[i]);
                dfs(i + 1, nums, path, res);
                path.pop_back();
            }
        }
    }
};

File: algorithms/90. Subsets II/solution.py
class Solution(object):
    def subsetsWithDup(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        nums.sort()
        res = [[]]
        for i in xrange(len(nums)):
            if i == 0 or nums[i] != nums[i-1]:
                l = len(res)
            for j in xrange(len(res) - l, len(res)):
                res.append(res[j] + [nums[i]])
        return res
